core/slice.go

```diff
@@ -817,16 +852,48 @@ func (sl *Slice) updatePhCache(pendingHeaderWithTermini types.PendingHeader, inS
 		}
 	}
 
+	termini := pendingHeaderWithTermini.Termini()
+
+	// This logic implements the Termini Carry
+	// * If a terminus exists in the cache, we can just copy the current value and make updates to it
+	// * If a terminus doesn't exist in the cache, that means that this is invoked
+	// 		on a dom block and we have to carry the parents Termini and make an update
+	var cachedTermini types.Termini
+	ph, exists := sl.readPhCache(pendingHeaderWithTermini.Termini().DomTerminus())
+	if exists {
+		cachedTermini = types.CopyTermini(ph.Termini())
+	} else {
+		parentHeader := sl.hc.GetHeaderOrCandidateByHash(pendingHeaderWithTermini.Header().ParentHash())
+		if parentHeader.Hash() == sl.config.GenesisHash {
+			ph, _ = sl.readPhCache(sl.config.GenesisHash)
+			cachedTermini = types.CopyTermini(ph.Termini())
+		} else {
+			localTermini := sl.hc.GetTerminiByHash(parentHeader.ParentHash())
+			ph, _ = sl.readPhCache(localTermini.DomTerminus())
+			cachedTermini = types.CopyTermini(ph.Termini())
+		}
+	}
+
+	// During the UpdateDom cycle we have to just copy what the Dom gives us
+	// otherwise update only the value that is being changed during inslice
+	// update and coord update
+	if !location.Equal(common.Location{}) {
+		cachedTermini.SetDomTerminiAtIndex(termini.DomTerminiAtIndex(location.DomIndex()), location.DomIndex())
+	} else {
+		for i := 0; i < len(termini.DomTermini()); i++ {
+			cachedTermini.SetDomTerminiAtIndex(termini.DomTerminiAtIndex(i), i)
+		}
+	}
+	cachedTermini.SetSubTermini(termini.SubTermini())
+
 	// Update the pendingHeader Cache
-	deepCopyPendingHeaderWithTermini := types.NewPendingHeader(types.CopyHeader(pendingHeaderWithTermini.Header()), pendingHeaderWithTermini.Termini())
+	deepCopyPendingHeaderWithTermini := types.NewPendingHeader(types.CopyHeader(pendingHeaderWithTermini.Header()), cachedTermini)
 	deepCopyPendingHeaderWithTermini.Header().SetLocation(common.NodeLocation)
 	deepCopyPendingHeaderWithTermini.Header().SetTime(uint64(time.Now().Unix()))
 
-	_, exists = sl.readPhCache(pendingHeaderWithTermini.Termini().DomTerminus())
-
 	if subReorg || !exists {
-		sl.writePhCache(pendingHeaderWithTermini.Termini().DomTerminus(), deepCopyPendingHeaderWithTermini)
-		log.Info("PhCache update:", "new terminus?:", !exists, "inSlice:", inSlice, "Ph Number:", deepCopyPendingHeaderWithTermini.Header().NumberArray(), "Termini:", deepCopyPendingHeaderWithTermini.Termini().DomTerminus())
+		sl.writePhCache(deepCopyPendingHeaderWithTermini.Termini().DomTerminus(), deepCopyPendingHeaderWithTermini)
+		log.Info("PhCache update:", "new terminus?:", !exists, "inSlice:", inSlice, "Ph Number:", deepCopyPendingHeaderWithTermini.Header().NumberArray(), "Termini:", deepCopyPendingHeaderWithTermini.Termini())
```

Comment:
we should use the .String method on Termini() here

Reply:
Not needed as its taken care by the String() function implemented on the Type Termini